### ISSUE TITLE
[didProvide take a lot of time] 

### ISSUE THREAD

**Issue by badredaha** [Adyen/adyen-ios#172]
**Describe the bug**
when i call didProvide sometimes work sometimes take a long time without a reponse !

**Comment by descorp** [comment]
Hi @badredaha 

Thank you for your feedback.

Standard timeout for HTTP requests is 30 seconds. 

Do you see a timeout or an error? Anything from your httpClient?
What do you see in your backend's logs?

---

Unless it is an error while processing Action response, unfortunately, there is nothing that could be done from the iOS SDK side.
Your app sends the request via your backend to a Checkout API, then we pass it to a chain of 3rd party APIs. In some extreme cases (3DS Fingerprint, for example), the response could take up to 5-10 seconds.

I can only suggest: 
* improve UX with "busy indicator" screen: show it on `didProvide` and hide when the response is handled;
*  increasing [priority of the network task](https://developer.apple.com/documentation/foundation/urlsessiontask/1410569-priority).

**Comment by badredaha** [comment]
sometime i have this error from didFail : error missingTransaction

**Comment by brurend** [comment]
Hi, I have a similar issue, it takes quite a while to go from the .handle(action) call until it gets to the didProvide, and during this time all I have is a blank white view showing.

Is this the intended behaviour or can I do anything about it?

**Comment by descorp** [comment]
Hi @brurend, 

Could you benchmark it (basic timer `print()`)? 
There are 5 steps in 3DS, would be helpful to know how long it takes for each:

1) Encrypted card data is send to your backend
2) `ThreeDS2FingerprintAction` received
3) `ActionComponentData` with 3DS Fingerprint is send to your backend
4) `ThreeDS2ChallengeAction` received
5) 3DS component visualised (this one would be harder to detect)

**Comment by brurend** [comment]
Hi @descorp ,

I tried to benchmark it, here are my results:

1. Encrypted card data sent at `2020-06-08 12:43:00.383431-0300`
Request time: 5.641668458

2. `ThreeDS2FingerprintAction` received at `2020-06-08 12:43:14.455648-0300`

3. `ActionComponentData` sent at `2020-06-08 12:43:16.132538-0300`
Request time: 3.848362667

4. `ThreeDS2ChallengeAction` received at `2020-06-08 12:43:19.983248-0300`

5. I couldn't get the exact time but its usually fast, I would say around 1 second or at most 1,5 seconds after step 4.

So in the end it took about 20 seconds to do everything, most of this was between step 1 and 2. The request to our backend did take a while, but after the response there were still 8~ seconds until I received the `FingerPrintAction` from the SDK.

**Comment by descorp** [comment]
Hi @brurend 

Good catch! 
We have similar results: the creation of ThreeDS2Fingerprint sometimes takes longer due to a Location service. It is a required step for 3DS 2 standard, however, location will be skipped if not accessible. 

Did you try the benchmark on a real device or a Simulator?
Could you check if the issue still presents if you disable the Location service?

In general: Location supposed to be gathered fast enough in real-life situations.

**Comment by descorp** [comment]
Hi @badredaha 

Are you still having the issue with `missingTransaction`?
If so, could you create another issue for it?

**Comment by brurend** [comment]
Hi @descorp 

Our app needs location services to be used to it should always be available during the process.

I have added a loading view during the creation of the Fingerprint and the flow is quite alright now, I just wanted to be sure there wasn't anything wrong with the loading time.

Thank you for the help.

**Comment by descorp** [comment]
Thanks @brurend !

I'll close this ticket for now then. If you or anyone still have concerns regarding 3DS timing - please, feel free to leave comments here or create another issue.
